### ISSUE TITLE
Messages & Read Receipts Repeating Infinitely

### ISSUE THREAD

**Issue by lanerussell** [vector-im/element-android#5690]
### Steps to reproduce

1. Where are you starting? What can you see?
    * Open app
    * Open bugged chat
    * Scroll back through message history
    * See duplicate messages and read receipts

You can see the "Encryption Enabled" message appears multiple times. This is the same for the few messages in the chat, and will repeat infinitely if I keep scrolling.
<details>
<summary>Screenshots</summary>

![screenshot](https://user-images.githubusercontent.com/20979879/161581410-6a79bda1-ae46-4754-9c23-83f471e2889a.png)

![screenshot2](https://user-images.githubusercontent.com/20979879/161581450-ca84b11b-1fee-476c-8175-69d509766d86.png)

</details>

#### What did you expect?
Only 1 copy of each message should appear, only 1 read receipt per user.

#### What happened instead?
Duplicates of both appear.

### Your phone model

Pixel 3

### Operating system version

Android 12

### Application version and app store

1.44 olm 3.2.10 Play Store

### Homeserver

matrix-synapse-py3/unknown,now 1.49.2+bionic1

### Will you send logs?

Yes

**Comment by ganfra** [comment]
Have you tried upgrading your app to the latest version?

**Comment by lanerussell** [comment]
My app has updated to 1.4.8 since this issue was opened, and the issue persists.

**Comment by lanerussell** [comment]
Sure thing. This is the only room I'm seeing this in. This is one of many of my rooms (encrypted and unencrypted) I upgraded to room version 8 around August 2021. I haven't been able to reproduce this bug by creating new rooms. It's interesting that the issue persists after removing the app and setting it back up from scratch.

I'll also include a screenshot from element-web which shows some additional older messages that don't appear on android. Web also shows the link to the old version of the room (which is empty at this point due to homeserver retention policy:
```
retention:
  enabled: true

  default_policy:
    min_lifetime: 1d
    max_lifetime: 90d

  allowed_lifetime_min: 1d
  allowed_lifetime_max: 90d

  purge_jobs:
    - longest_max_lifetime: 3d
      interval: 1h
    - shortest_max_lifetime: 3d
      interval: 1d
```

<details>
<summary>Element Web Screenshot</summary>

![1](https://user-images.githubusercontent.com/20979879/162993819-9d31be76-d7ff-4dec-b879-f3d935c678fd.png)
</details>

**Comment by mirenbz** [comment]
This is happening to me as well. Only in the Android app; desktop and web look fine. Not in all rooms either: one v4, one v5 and one v6, but not another v4 room 🤷🏼

**Comment by weeman1337** [comment]
This may be a related rageshake: https://github.com/matrix-org/element-android-rageshakes/issues/39315

**Comment by JayFoxRox** [comment]
I'm also affected on 1.4.20.

I think I know why this problem exists though. The problematic chat in my case is on hackint IRC https://hackint.org/transport/matrix.
Their retention policy is to only send you messages up until 2 minutes ago since join.

The web client shows "You can't see earlier messages" when scrolling up.
The android client, however, tries to loads non-existing messages as soon as I scroll up.

Because there are no earlier messages than what's already downloaded, it keeps adding the same handful of messages. Each time a block of messages is loaded, the "Seen by" counters also increment.

The timeline looks unordered / skewed as a result.

Interestingly, the android client loads messages even further ago than 2 minutes (for whatever reason).
On the web version I only see events from 28th of July (today).
On the Android version I see more events (although exclusively join/leave messages, and not chat messages before the 28th of July) this pattern repeating (read from bottom to top, as if this was the message window):

- [Swirl, which will load messages from the 28th July again]
- April 6
- April 7
- June 26
- June 27
- June 28
- 28th July [second repeat]
- April 6
- April 7
- June 26
- June 27
- June 28
- 28th July [first repeat]
- April 6
- April 7
- June 26
- June 27
- June 28
- 28th July [newest message]

So I believe this happens because the android client does not respect the boundary / "no more earlier messages".
Each time it load the same message again, you also seem to get the read receipt again.

Furthermore, the retention for join/leave messages appears to be broken (likely due to the member list being reconstructed from join/leave messages, so they are replayed? - the web client does not show this though; the android version does)

**Comment by jmartinesp** [comment]
This should have been fixed by #6318, please re-open the issue otherwise.
